File: src/character.ts
import { DragonElement } from './dragon.js';

/**
 * Character traits that determine how a dragon behaves and interacts
 * Values range from 0 to 100
 */
export interface CharacterTraits {
  /** How friendly and approachable the dragon is (0-100) */
  friendliness: number;
  /** How aggressive or confrontational the dragon is (0-100) */
  aggression: number;
  /** How much the dragon enjoys social interactions (0-100) */
  sociability: number;
  /** How curious and exploratory the dragon is (0-100) */
  curiosity: number;
  /** How playful the dragon is (0-100) */
  playfulness: number;
  /** How dominant or submissive the dragon is (0-100, higher = more dominant) */
  dominance: number;
  /** How patient the dragon is (0-100) */
  patience: number;
}

/**
 * Preferences that influence how dragons respond to others
 */
export interface CharacterPreferences {
  /** Elements this dragon prefers (positive interactions) */
  preferredElements: DragonElement[];
  /** Elements this dragon dislikes (negative interactions) */
  dislikedElements: DragonElement[];
  /** Personality traits this dragon is attracted to */
  preferredTraits: (keyof CharacterTraits)[];
  /** Personality traits this dragon avoids */
  dislikedTraits: (keyof CharacterTraits)[];
}

/**
 * Complete character profile for a dragon
 */
export class DragonCharacter {
  public traits: CharacterTraits;
  public preferences: CharacterPreferences;

  constructor(traits?: Partial<CharacterTraits>, preferences?: Partial<CharacterPreferences>) {
    this.traits = {
      friendliness: traits?.friendliness ?? this.randomTrait(),
      aggression: traits?.aggression ?? this.randomTrait(),
      sociability: traits?.sociability ?? this.randomTrait(),
      curiosity: traits?.curiosity ?? this.randomTrait(),
      playfulness: traits?.playfulness ?? this.randomTrait(),
      dominance: traits?.dominance ?? this.randomTrait(),
      patience: traits?.patience ?? this.randomTrait(),
    };

    this.preferences = {
      preferredElements: preferences?.preferredElements ?? [],
      dislikedElements: preferences?.dislikedElements ?? [],
      preferredTraits: preferences?.preferredTraits ?? [],
      dislikedTraits: preferences?.dislikedTraits ?? [],
    };
  }

  /**
   * Generates a random trait value (0-100)
   * Uses a bell curve distribution (most values around 50, extremes less common)
   */
  private randomTrait(): number {
    // Generate two random numbers and average them for a bell curve effect
    const value1 = Math.random() * 100;
    const value2 = Math.random() * 100;
    return Math.round((value1 + value2) / 2);
  }

  /**
   * Calculates compatibility score with another dragon's character
   * Returns a value from -100 (very incompatible) to 100 (very compatible)
   */
  public getCompatibilityWith(other: DragonCharacter, otherElement: DragonElement): number {
    let score = 0;

    // Trait compatibility
    const traitPairs: Array<[keyof CharacterTraits, keyof CharacterTraits]> = [
      ['friendliness', 'friendliness'],
      ['sociability', 'sociability'],
      ['playfulness', 'playfulness'],
      ['patience', 'patience'],
    ];

    for (const [trait1, trait2] of traitPairs) {
      const diff = Math.abs(this.traits[trait1] - other.traits[trait2]);
      score += (100 - diff) / 10; // Closer values = higher compatibility
    }

    // Dominance compatibility (opposites attract, but similar can work too)
    const dominanceDiff = Math.abs(this.traits.dominance - other.traits.dominance);
    if (dominanceDiff < 30) {
      score += 10; // Similar dominance levels work well
    } else if (dominanceDiff > 70) {
      score += 15; // Very different can also work (complementary)
    }

    // Aggression compatibility (low aggression pairs well with most)
    if (this.traits.aggression < 30 && other.traits.aggression < 30) {
      score += 20; // Both peaceful
    } else if (this.traits.aggression > 70 && other.traits.aggression > 70) {
      score -= 30; // Both aggressive = conflict
    }

    // Element preferences
    if (this.preferences.preferredElements.includes(otherElement)) {
      score += 25;
    }
    if (this.preferences.dislikedElements.includes(otherElement)) {
      score -= 30;
    }

    // Trait preferences
    for (const preferredTrait of this.preferences.preferredTraits) {
      if (other.traits[preferredTrait] > 60) {
        score += 10;
      }
    }

    for (const dislikedTrait of this.preferences.dislikedTraits) {
      if (other.traits[dislikedTrait] > 60) {
        score -= 15;
      }
    }

    // Normalize to -100 to 100 range
    return Math.max(-100, Math.min(100, score));
  }

  /**
   * Determines the interaction style based on traits
   */
  public getInteractionStyle(): 'friendly' | 'playful' | 'serious' | 'aggressive' | 'shy' | 'curious' {
    if (this.traits.aggression > 70) return 'aggressive';
    if (this.traits.friendliness > 70 && this.traits.playfulness > 60) return 'playful';
    if (this.traits.friendliness > 70) return 'friendly';
    if (this.traits.sociability < 30) return 'shy';
    if (this.traits.curiosity > 70) return 'curious';
    return 'serious';
  }
}

/**
 * Generates a random character for a dragon
 */
export function generateRandomCharacter(element?: DragonElement): DragonCharacter {
  const traits: Partial<CharacterTraits> = {};
  const preferences: Partial<CharacterPreferences> = {
    preferredElements: [],
    dislikedElements: [],
    preferredTraits: [],
    dislikedTraits: [],
  };

  // Element-based character adjustments
  if (element) {
    switch (element) {
      case 'Fire':
        traits.aggression = Math.min(100, (traits.aggression ?? 50) + 20);
        traits.dominance = Math.min(100, (traits.dominance ?? 50) + 15);
        preferences.preferredElements = ['Fire', 'Lightning'];
        preferences.dislikedElements = ['Water', 'Ice'];
        break;
      case 'Water':
        traits.patience = Math.min(100, (traits.patience ?? 50) + 20);
        traits.friendliness = Math.min(100, (traits.friendliness ?? 50) + 15);
        preferences.preferredElements = ['Water', 'Ice'];
        preferences.dislikedElements = ['Fire', 'Lightning'];
        break;
      case 'Earth':
        traits.patience = Math.min(100, (traits.patience ?? 50) + 25);
        traits.curiosity = Math.max(0, (traits.curiosity ?? 50) - 15);
        preferences.preferredElements = ['Earth', 'Wind'];
        break;
      case 'Wind':
        traits.curiosity = Math.min(100, (traits.curiosity ?? 50) + 20);
        traits.playfulness = Math.min(100, (traits.playfulness ?? 50) + 15);
        preferences.preferredElements = ['Wind', 'Lightning'];
        break;
      case 'Lightning':
        traits.aggression = Math.min(100, (traits.aggression ?? 50) + 15);
        traits.curiosity = Math.min(100, (traits.curiosity ?? 50) + 20);
        preferences.preferredElements = ['Lightning', 'Fire'];
        preferences.dislikedElements = ['Earth'];
        break;
      case 'Ice':
        traits.sociability = Math.max(0, (traits.sociability ?? 50) - 20);
        traits.patience = Math.min(100, (traits.patience ?? 50) + 25);
        preferences.preferredElements = ['Ice', 'Water'];
        preferences.dislikedElements = ['Fire'];
        break;
    }
  }

  return new DragonCharacter(traits, preferences);
}

File: rust/src/character.rs
use crate::dragon::DragonElement;
use crate::values::{DragonValues, generate_dragon_values};

#[derive(Debug, Clone, Copy)]
pub struct CharacterTraits {
    // Non-conflicting individual traits (0-100)
    
    /// Warmth and openness toward others
    /// Low (0-30): Reserved, cautious, or standoffish
    /// Medium (31-70): Generally approachable, balanced social behavior
    /// High (71-100): Warm, welcoming, and open to others
    pub friendliness: u32,
    
    /// Preference for social interaction vs solitude
    /// Low (0-30): Prefers solitude, introverted, shy
    /// Medium (31-70): Enjoys some social time, balanced
    /// High (71-100): Thrives in groups, extroverted, seeks company
    pub sociability: u32,
    
    /// Interest in exploring and discovering new things
    /// Low (0-30): Content with familiar things, less exploratory
    /// Medium (31-70): Moderate interest in new experiences
    /// High (71-100): Eager to explore, discover, and learn
    pub curiosity: u32,
    
    /// Tendency toward fun and lightheartedness vs seriousness
    /// Low (0-30): Serious, focused, business-like
    /// Medium (31-70): Can have fun but also serious when needed
    /// High (71-100): Fun-loving, energetic, enjoys games and activities
    pub playfulness: u32,
    
    /// Natural leadership and assertiveness
    /// Low (0-30): Submissive, follows others, prefers not to lead
    /// Medium (31-70): Balanced leadership, can lead or follow
    /// High (71-100): Natural leader, takes charge, assertive
    pub dominance: u32,
    
    // Conflicting pairs represented as axes (0-100)
    // Higher value = more of the first trait, lower = more of the second
    
    /// Aggression (high) vs Patience (low) axis
    /// 0 = max patience (very tolerant, calm, takes time to consider)
    /// 100 = max aggression (confrontational, competitive, quick to challenge)
    /// These traits conflict: you cannot be both quick to challenge and take time to consider
    pub aggression_vs_patience: u32,
}

#[derive(Debug, Clone)]
pub struct DragonCharacter {
    pub traits: CharacterTraits,
    pub values: DragonValues,
}

impl CharacterTraits {
    /// Get individual trait from an axis
    /// For axes where higher = first trait, lower = second trait
    fn get_first_from_axis(axis: u32) -> u32 {
        axis
    }
    
    fn get_second_from_axis(axis: u32) -> u32 {
        100u32.saturating_sub(axis)
    }
    
    // Convenience getters for individual traits (for backward compatibility)
    pub fn aggression(&self) -> u32 {
        Self::get_first_from_axis(self.aggression_vs_patience)
    }
    
    pub fn patience(&self) -> u32 {
        Self::get_second_from_axis(self.aggression_vs_patience)
    }
}

impl DragonCharacter {
    pub fn new(
        traits: Option<CharacterTraits>,
        values: Option<DragonValues>,
    ) -> Self {
        DragonCharacter {
            traits: traits.unwrap_or_else(|| CharacterTraits {
                friendliness: random_trait(),
                sociability: random_trait(),
                curiosity: random_trait(),
                playfulness: random_trait(),
                dominance: random_trait(),
                aggression_vs_patience: random_trait(),
            }),
            values: values.unwrap_or_else(|| generate_dragon_values(None)),
        }
    }

    pub fn get_interaction_style(&self) -> String {
        if self.traits.aggression() > 70 {
            "aggressive".to_string()
        } else if self.traits.friendliness > 70 && self.traits.playfulness > 60 {
            "playful".to_string()
        } else if self.traits.friendliness > 70 {
            "friendly".to_string()
        } else if self.traits.sociability < 30 {
            "shy".to_string()
        } else if self.traits.curiosity > 70 {
            "curious".to_string()
        } else {
            "serious".to_string()
        }
    }

    pub fn traits(&self) -> &CharacterTraits {
        &self.traits
    }

    pub fn values(&self) -> &DragonValues {
        &self.values
    }
}

fn random_trait() -> u32 {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    let value1 = rng.gen_range(0..=100);
    let value2 = rng.gen_range(0..=100);
    ((value1 + value2) / 2) as u32
}

pub fn generate_random_character(element: Option<DragonElement>) -> DragonCharacter {
    // Generate random axis values (0-100)
    // For axes: higher = first trait, lower = second trait
    let mut traits = CharacterTraits {
        friendliness: random_trait(),
        sociability: random_trait(),
        curiosity: random_trait(),
        playfulness: random_trait(),
        dominance: random_trait(),
        aggression_vs_patience: random_trait(),
    };

    let values = generate_dragon_values(element);

    // Element-based character adjustments
    // Adjust axes by moving toward the preferred trait
    if let Some(element) = element {
        match element {
            DragonElement::Fire => {
                // Higher aggression, higher dominance
                traits.aggression_vs_patience = (traits.aggression_vs_patience + 20).min(100);
                traits.dominance = (traits.dominance + 15).min(100);
            }
            DragonElement::Water => {
                // Higher patience, higher friendliness
                traits.aggression_vs_patience = traits.aggression_vs_patience.saturating_sub(20);
                traits.friendliness = (traits.friendliness + 15).min(100);
            }
            DragonElement::Earth => {
                // Higher patience, lower curiosity
                traits.aggression_vs_patience = traits.aggression_vs_patience.saturating_sub(25);
                traits.curiosity = traits.curiosity.saturating_sub(15);
            }
            DragonElement::Wind => {
                // Higher curiosity, higher playfulness
                traits.curiosity = (traits.curiosity + 20).min(100);
                traits.playfulness = (traits.playfulness + 15).min(100);
            }
            DragonElement::Lightning => {
                // Higher aggression, higher curiosity
                traits.aggression_vs_patience = (traits.aggression_vs_patience + 15).min(100);
                traits.curiosity = (traits.curiosity + 20).min(100);
            }
            DragonElement::Ice => {
                // Lower sociability, higher patience
                traits.sociability = traits.sociability.saturating_sub(20);
                traits.aggression_vs_patience = traits.aggression_vs_patience.saturating_sub(25);
            }
        }
    }

    // Ensure all values are in valid range
    traits.friendliness = traits.friendliness.min(100);
    traits.sociability = traits.sociability.min(100);
    traits.curiosity = traits.curiosity.min(100);
    traits.playfulness = traits.playfulness.min(100);
    traits.dominance = traits.dominance.min(100);
    traits.aggression_vs_patience = traits.aggression_vs_patience.min(100);

    DragonCharacter::new(Some(traits), Some(values))
}
